File: hash.go
// hash.go
package main

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"encoding/hex"
	"flag"
	"fmt"
	"io"
	"os"
	"strings"
)

// ANSI colors
const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type hashFunc func(data []byte) []byte

func getHasher(algo string) (hashFunc, error) {
	switch algo {
	case "md5":
		return func(data []byte) []byte {
			h := md5.Sum(data)
			return h[:]
		}, nil
	case "sha1":
		return func(data []byte) []byte {
			h := sha1.Sum(data)
			return h[:]
		}, nil
	case "sha256":
		return func(data []byte) []byte {
			h := sha256.Sum256(data)
			return h[:]
		}, nil
	case "sha512":
		return func(data []byte) []byte {
			h := sha512.Sum512(data)
			return h[:]
		}, nil
	default:
		return nil, fmt.Errorf("unsupported algorithm: %s", algo)
	}
}

func formatHash(digest []byte, format string) string {
	if format == "hex" {
		return hex.EncodeToString(digest)
	} else if format == "base64" {
		return base64.StdEncoding.EncodeToString(digest)
	}
	return hex.EncodeToString(digest) // fallback
}

func readFile(filename string) ([]byte, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	return io.ReadAll(file)
}

func main() {
	algo := flag.String("a", "sha256", "Algorithm: md5, sha1, sha256, sha512")
	text := flag.String("t", "", "Text to hash")
	input := flag.String("i", "", "Input file")
	output := flag.String("o", "", "Output file")
	salt := flag.String("s", "", "Salt string")
	compare := flag.String("c", "", "Hash to compare against")
	format := flag.String("f", "hex", "Output format: hex, base64")
	help := flag.Bool("h", false, "Show help")
	flag.Parse()

	if *help {
		fmt.Println(`Usage: hash [options]
Options:
  -a <algo>     Algorithm: md5, sha1, sha256, sha512 (default sha256)
  -t <text>     Text to hash
  -i <file>     Input file
  -o <file>     Output file
  -s <salt>     Salt string
  -c <hash>     Hash to compare against
  -f <format>   Output format: hex, base64 (default hex)
  -h            Show this help`)
		return
	}

	// Read input
	var data []byte
	if *text != "" {
		data = []byte(*text)
	} else if *input != "" {
		d, err := readFile(*input)
		if err != nil {
			fmt.Println(colorize("Error reading file: "+err.Error(), red))
			os.Exit(1)
		}
		data = d
	} else {
		// read from stdin
		stat, _ := os.Stdin.Stat()
		if (stat.Mode() & os.ModeCharDevice) == 0 {
			// data from pipe
			d, err := io.ReadAll(os.Stdin)
			if err != nil {
				fmt.Println(colorize("Error reading stdin: "+err.Error(), red))
				os.Exit(1)
			}
			data = d
		} else {
			fmt.Print(colorize("Enter data (Ctrl+D to finish): ", yellow))
			d, err := io.ReadAll(os.Stdin)
			if err != nil {
				fmt.Println(colorize("Error reading stdin: "+err.Error(), red))
				os.Exit(1)
			}
			data = d
		}
	}
	if len(data) == 0 {
		fmt.Println(colorize("No input data", red))
		os.Exit(1)
	}

	// Get hasher
	hasher, err := getHasher(*algo)
	if err != nil {
		fmt.Println(colorize(err.Error(), red))
		os.Exit(1)
	}

	// Prepend salt if given
	finalData := data
	if *salt != "" {
		finalData = append([]byte(*salt), data...)
	}

	// Compute hash
	digest := hasher(finalData)
	result := formatHash(digest, *format)

	// Compare if requested
	if *compare != "" {
		if result == *compare {
			fmt.Println(colorize("✅ Hash matches!", green))
		} else {
			fmt.Println(colorize("❌ Hash does NOT match.", red))
			fmt.Printf("Expected: %s\n", *compare)
			fmt.Printf("Computed: %s\n", result)
		}
		os.Exit(0)
	}

	// Output
	if *output != "" {
		err := os.WriteFile(*output, []byte(result), 0644)
		if err != nil {
			fmt.Println(colorize("Error writing file: "+err.Error(), red))
			os.Exit(1)
		}
		fmt.Println(colorize("Hash written to "+*output, green))
	} else {
		fmt.Println(result)
	}
}

File: hash.js
// hash.js
#!/usr/bin/env node
'use strict';

const crypto = require('crypto');
const fs = require('fs');
const readline = require('readline');

// ANSI colors
const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

function getHasher(algo) {
    const algos = ['md5', 'sha1', 'sha256', 'sha512'];
    if (!algos.includes(algo)) {
        throw new Error(`Unsupported algorithm: ${algo}`);
    }
    return (data) => {
        const hash = crypto.createHash(algo);
        hash.update(data);
        return hash.digest();
    };
}

function formatHash(digest, format) {
    if (format === 'hex') return digest.toString('hex');
    if (format === 'base64') return digest.toString('base64');
    return digest.toString('hex');
}

function readFile(filename) {
    return fs.readFileSync(filename);
}

function writeFile(filename, content) {
    fs.writeFileSync(filename, content, 'utf8');
}

function parseArgs() {
    const args = process.argv.slice(2);
    const opts = {
        algo: 'sha256',
        text: '',
        input: '',
        output: '',
        salt: '',
        compare: '',
        format: 'hex',
        help: false
    };
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        switch (arg) {
            case '-a':
            case '--algo': opts.algo = args[++i]; break;
            case '-t':
            case '--text': opts.text = args[++i]; break;
            case '-i':
            case '--input': opts.input = args[++i]; break;
            case '-o':
            case '--output': opts.output = args[++i]; break;
            case '-s':
            case '--salt': opts.salt = args[++i]; break;
            case '-c':
            case '--compare': opts.compare = args[++i]; break;
            case '-f':
            case '--format': opts.format = args[++i]; break;
            case '-h':
            case '--help': opts.help = true; break;
            default:
                console.error(colorize(`Unknown option: ${arg}`, 'red'));
                process.exit(1);
        }
    }
    return opts;
}

async function main() {
    const opts = parseArgs();
    if (opts.help) {
        console.log(`Usage: node hash.js [options]
Options:
  -a, --algo <algo>   Algorithm: md5, sha1, sha256, sha512 (default sha256)
  -t, --text <text>   Text to hash
  -i, --input <file>  Input file
  -o, --output <file> Output file
  -s, --salt <salt>   Salt string
  -c, --compare <hash> Hash to compare against
  -f, --format <fmt>  Output format: hex, base64 (default hex)
  -h, --help          Show this help`);
        return;
    }

    // Read input
    let data;
    if (opts.text) {
        data = Buffer.from(opts.text, 'utf8');
    } else if (opts.input) {
        try {
            data = readFile(opts.input);
        } catch (err) {
            console.error(colorize(`Error reading file: ${err.message}`, 'red'));
            process.exit(1);
        }
    } else {
        // read from stdin
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: false
        });
        const chunks = [];
        for await (const chunk of rl) {
            chunks.push(Buffer.from(chunk + '\n', 'utf8'));
        }
        if (chunks.length === 0) {
            console.error(colorize('No input data', 'red'));
            process.exit(1);
        }
        data = Buffer.concat(chunks);
        // remove trailing newline if present? We'll keep as is.
    }

    // Prepare data with salt
    let finalData = data;
    if (opts.salt) {
        finalData = Buffer.concat([Buffer.from(opts.salt, 'utf8'), data]);
    }

    // Compute hash
    let hasher;
    try {
        hasher = getHasher(opts.algo);
    } catch (err) {
        console.error(colorize(err.message, 'red'));
        process.exit(1);
    }
    const digest = hasher(finalData);
    const result = formatHash(digest, opts.format);

    // Compare
    if (opts.compare) {
        if (result === opts.compare) {
            console.log(colorize('✅ Hash matches!', 'green'));
        } else {
            console.log(colorize('❌ Hash does NOT match.', 'red'));
            console.log(`Expected: ${opts.compare}`);
            console.log(`Computed: ${result}`);
        }
        process.exit(0);
    }

    // Output
    if (opts.output) {
        try {
            writeFile(opts.output, result);
            console.log(colorize(`Hash written to ${opts.output}`, 'green'));
        } catch (err) {
            console.error(colorize(`Error writing file: ${err.message}`, 'red'));
            process.exit(1);
        }
    } else {
        console.log(result);
    }
}

main().catch(err => {
    console.error(colorize(`Unhandled error: ${err.message}`, 'red'));
    process.exit(1);
});
